File: flutter/photo_gallery_proto/lib/auth_repository.dart
import 'dart:async';
import 'package:amplify_auth_cognito/amplify_auth_cognito.dart';
import 'package:amplify_flutter/amplify.dart';

class AuthRepository {
  StreamSubscription observeAuth() {
    return Amplify.Hub.listen([HubChannel.Auth], (event) {
      print(event);
    });
  }

  Future<void> signUp({String username, String email, String password}) async {
    try {
      final options = SignUpOptions(
        userAttributes: {'email': email},
      );
      await Amplify.Auth.signUp(
        username: username,
        password: password,
        options: options,
      );
    } catch (e) {
      throw e;
    }
  }

  Future<void> login({String username, String password}) async {
    try {
      await Amplify.Auth.signIn(
        username: username,
        password: password,
      );
    } catch (e) {
      throw e;
    }
  }

  Future<void> confirmSignUp({String username, String confirmationCode}) async {
    try {
      await Amplify.Auth.confirmSignUp(
        username: username,
        confirmationCode: confirmationCode,
      );
    } catch (e) {
      throw e;
    }
  }
}

File: android/auth-flow-proto/app/src/main/java/com/example/auth_flow_proto/AuthRepository.kt
package com.example.auth_flow_proto

import android.util.Log
import com.amplifyframework.auth.AuthUserAttributeKey
import com.amplifyframework.auth.options.AuthSignUpOptions
import com.amplifyframework.core.Amplify
import java.lang.Exception

class AuthRepository {
    fun fetchAuthSession(session: (Boolean) -> Unit) {
        Amplify.Auth.fetchAuthSession(
            { session(it.isSignedIn) },
            { Log.e("kilo", "Failed to fetch auth session", it) }
        )
    }

    fun signUp(username: String, email: String, password: String, onCompletion: (Exception?) -> Unit) {
        val options = AuthSignUpOptions.builder()
            .userAttribute(AuthUserAttributeKey.email(), email)
            .build()

        Amplify.Auth.signUp(username, password, options,
            { onCompletion(null) },
            { onCompletion(it) }
        )
    }

    fun confirmSignUp(username: String, confirmationCode: String, onCompletion: (Exception?) -> Unit) {
        Amplify.Auth.confirmSignUp(username, confirmationCode,
            { onCompletion(null) },
            { onCompletion(it) }
        )
    }

    fun login(username: String, password: String, onCompletion: (Exception?) -> Unit) {
        Amplify.Auth.signIn(username, password,
            { onCompletion(null) },
            { onCompletion(it) }
        )
    }

    fun signOut(onCompletion: (Exception?) -> Unit) {
        Amplify.Auth.signOut(
            { onCompletion(null) },
            { onCompletion(it) }
        )
    }
}